File: 121. Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        
        if(prices.size() < 2) return 0;
        
        vector<int> maxProfit(prices.size());
        maxProfit[0] = 0;
        
        int tempMin = prices[0];
        
        vector<int>::iterator a = prices.begin() + 1;
        
        int i = 1;
        
        for(; a != prices.end(); ++a){
            
            tempMin = min(tempMin, *(a - 1));
            
            maxProfit[i] = max(maxProfit[i - 1], *a - tempMin);
            
            ++i;
            
        }
        
        return maxProfit[i - 1];
        
    }
};

File: 121_Best_Time_to_Buy_and_Sell_Stock.java
class Solution {

    public int maxProfit(int[] prices) {
        
        /* Sanity check */
        if(prices.length == 0)
            return 0;

        int min_buy = prices[0];        /* Minimal buy-in price */
        int max_profit = 0;             /* Max profit */

        for(int i = 1; i < prices.length; ++i) {
            int profit = prices[i] - min_buy;
            if(profit > max_profit)
                max_profit = profit;
            if(prices[i] < min_buy)
                min_buy = prices[i];
        }

        return max_profit;
    }
}